scheds/rust/scx_rustlite/src/main.rs

```diff
@@ -0,0 +1,480 @@
+            min_vruntime,
+            nr_cpus_online,
+            struct_ops,
+        })
+    }
+
+    // Read exit code from the BPF part.
+    fn read_bpf_exit_kind(&mut self) -> i32 {
+        unsafe { std::ptr::read_volatile(&self.skel.bss().exit_kind as *const _) }
+    }
+
+    // Called on exit to get exit code and exit message from the BPF part.
+    fn report_bpf_exit_kind(&mut self) -> Result<()> {
+        match self.read_bpf_exit_kind() {
+            0 => Ok(()),
+            etype => {
+                let cstr = unsafe { CStr::from_ptr(self.skel.bss().exit_msg.as_ptr() as *const _) };
+                let msg = cstr
+                    .to_str()
+                    .context("Failed to convert exit msg to string")
+                    .unwrap();
+                info!("BPF exit_kind={} msg={}", etype, msg);
+                Ok(())
+            }
+        }
+    }
+
+    // Update task's vruntime based on the information collected from the kernel part.
+    fn update_enqueued(
+        task_info: &mut TaskInfo,
+        sum_exec_runtime: u64,
+        weight: u64,
+        min_vruntime: u64,
+    ) {
+        // Add cputime delta normalized by weight to the vruntime (if delta > 0).
+        if sum_exec_runtime > task_info.sum_exec_runtime {
+            let delta = (sum_exec_runtime - task_info.sum_exec_runtime) / weight;
+            task_info.vruntime += delta;
+        }
+        // Make sure vruntime is moving forward (> current minimum).
+        if min_vruntime > task_info.vruntime {
+            task_info.vruntime = min_vruntime;
+        }
+        // Update total task cputime.
+        task_info.sum_exec_runtime = sum_exec_runtime;
+    }
+
+    // Drain all the tasks from the queued list, update their vruntime (Self::update_enqueued()),
+    // then push them all to the task pool (doing so will sort them by their vruntime).
+    fn drain_queued_tasks(&mut self) {
+        let maps = self.skel.maps();
+        let queued = maps.queued();
+
+        loop {
+            match queued.lookup_and_delete(&[]) {
+                Ok(Some(msg)) => {
+                    // Schedule the task and update task information.
+                    let task = EnqueuedMessage::from_bytes(msg.as_slice()).as_queued_task_ctx();
+                    if let Some(task_info) = self.task_map.get_mut(task.pid) {
+                        Self::update_enqueued(
+                            task_info,
+                            task.sum_exec_runtime,
+                            task.weight,
+                            self.min_vruntime,
+                        );
+                        self.task_pool.push(task.pid, task_info.vruntime);
+                    } else {
+                        let task_info = TaskInfo {
+                            sum_exec_runtime: task.sum_exec_runtime,
+                            vruntime: self.min_vruntime,
+                        };
+                        self.task_map.insert(task.pid, task_info);
```

Comment:
Another non-blocker: We probably wanna make the BPF side send "task exited" message to clean up the userside context. A tricky thing here is that `.disable()` callback can't wait so the signalling queue can overflow.

An imperfect solution which works most of the time should be fine here but the canonical solution would be establishing a common mmap'd map entry between user and BPF code during `.prep_enable()` and indicate task exit by chaining them from BPF side so that there's no memory allocation involved in signalling exit. Might be another thing which would be nice to have in scx_utils.

Reply:
... or the scheduler could implement its own garbage collector to figure out the pids that are not running anymore, but having a feedback from the BPF part would be nicer. I'll add a TODO note for this in the code.